nemo/collections/asr/losses/pt_losses/contrastive.py

```diff
@@ -0,0 +1,206 @@
+# ! /usr/bin/python
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from nemo.core import Loss, typecheck
+from nemo.core.neural_types import LossType, NeuralType, SpectrogramType, VoidType
+
+__all__ = ["ContrastiveLoss"]
+
+
+def buffered_arange(max):
+    if not hasattr(buffered_arange, "buf"):
+        buffered_arange.buf = torch.LongTensor()
+    if max > buffered_arange.buf.numel():
+        buffered_arange.buf.resize_(max)
+        torch.arange(max, out=buffered_arange.buf)
+    return buffered_arange.buf[:max]
+
+
+class ContrastiveLoss(Loss):
+    @property
+    def input_types(self):
+        """Input types definitions for Contrastive.
+        """
+        return {
+            "spectrograms": NeuralType(("B", "D", "T"), SpectrogramType()),
+            "spec_masks": NeuralType(("B", "D", "T"), SpectrogramType()),
+            "decoder_outputs": NeuralType(("B", "T", "D"), VoidType()),
+        }
+
+    @property
+    def output_types(self):
+        """Output types definitions for Contrastive.
+        loss:
+            NeuralType(None)
+        """
+        return {"loss": NeuralType(elements_type=LossType())}
+
+    def __init__(
+        self,
+        in_dim,
+        proj_dim=128,
+        combine_time_steps=1,
+        n_negatives=100,
+        quantized_targets=True,
+        codebook_size=320,
+        prob_ppl_weight=0.1,
+        logit_temp=0.1,
+        reduce=True,
+        sample_from_non_masked=True,
+        sample_from_codebook=False,
+        group_loss=False,
+        num_groups=2,
+        temp=(2, 0.5, 0.999995),
```

Comment:
Tuples are safer than lists, but default to None and create the tuple inside init.